src/ProphecyTrait.php

```diff
@@ -36,7 +36,10 @@ trait ProphecyTrait
      * @throws DoubleException
      * @throws InterfaceNotFoundException
      *
-     * @psalm-param class-string|null $classOrInterface
+     * @template T of object
+     * @phpstan-param class-string<T>|null $classOrInterface
+     * @phpstan-return ($classOrInterface is null ? ObjectProphecy<object> : ObjectProphecy<T>)
```

Comment:
Both phpstan and psalm will read the annotation with the prefix of the other tool if they don't have their own prefixed annotation (but with a different error handling as they will skip an annotation they cannot support instead of reporting an error).
The reason they do that is to avoid forcing all libraries to provide duplicate annotations to let downstream projects benefit from their type when being analyzed.

The general rule is that libraries should apply the prefix corresponding to the tool they use themselves (so that their own analysis report invalid annotations). And as I'm more a phpstan user than a psalm one myself, I would be using phpstan when adding a static analysis setup in this package (which I will probably do at some point).

I'm not using `@return` here because conditional types are supported by psalm and phpstan but not by other tools dealing with phpdoc, which might then be confused about them.

Reply:
Thank you for the explanation! Obviously I wasn't aware on how they approach this but it makes sense as writing two types of anotations would be a maintenance burden.